cisc/certificate.go

```diff
@@ -0,0 +1,329 @@
+	_, err := pemToCertificate(cert)
+	return err == nil
+}
+
+// Check if the string in argument correspond to a valid certificate
+// Return true if it is the case and false otherwise
+func isValid(cert string) bool {
+	rootPool := x509.NewCertPool()
+	certRoots, err := pemToCertificate(LetsencryptCert)
+	if err != nil {
+		log.Info("The root certificate is not valid")
+		return false
+	}
+	rootPool.AddCert(certRoots)
+	certCheck, err := pemToCertificate(cert)
+	if err != nil {
+		log.Info("The checked certificate is not valid")
+		return false
+	}
+	opts := x509.VerifyOptions{
+		DNSName: certCheck.DNSNames[0],
+		Roots:   rootPool,
+	}
+	if _, err := certCheck.Verify(opts); err != nil {
+		log.Info("The certificate is not valid: " + err.Error())
+		return false
+	}
+	log.Info("Certificate OK")
+	return true
+}
+
+// Renew the certificate that is caracterized by its string version
+// and returns the new certificate
+func renewCert(cert string) (string, error) {
+
+	oldcert, err := pemToCertificate(cert)
+	if err != nil {
+		return "", err
+	}
+
+	// Get the serial number and compute the URI
+	serial := fmt.Sprintf("%x", oldcert.SerialNumber)
+	certuri := "https://acme-v01.api.letsencrypt.org/acme/cert/0" + serial
+
+	// Create a client to the acme of letsencrypt.
+	cli, err := letsencrypt.NewClient(LetsEncryptURL)
+
+	if err != nil {
+		return "", err
+	}
+
+	// Renew the certificate
+	certRen, err := cli.RenewCertificate(string(certuri))
+	if err != nil {
+		return "", err
+	}
+
+	// Write in the memory
+	fullchain, err := cli.Bundle(certRen)
+	if err != nil {
+		return "", errors.New("can't bundle the certificate" + err.Error())
+	}
+	ioutil.WriteFile("fullchain.pem", fullchain, 0644)
+	log.Info("Certificate successfully renewed")
+
+	return string(fullchain), nil
+}
+
+// Revoke a certificate by giving as arguments the certificate and the path to
+// the registerkey.pem key. This key is created when registering to the ACME
+// server
+func revokeCert(path string, cert string) error {
+
+	// Create a client to the acme of letsencrypt
+	log.Info("Revoking the key")
+	cli, err := letsencrypt.NewClient(LetsEncryptURL)
+
+	if err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return errors.New("key can not be found")
+		}
+	}
+
+	// Find or generate the private key and register with the ACME server.
+	var key *rsa.PrivateKey
+	key, err = loadKey(path)
+	if err != nil {
+		return err
+	}
+
+	// Revoke the certificate
+	err = cli.RevokeCertificate(key, []byte(cert))
+	if err != nil {
+		return err
+	}
+	log.Info("Succesfully revoked")
+	return nil
+}
+
+// Code adapted from: https://ericchiang.github.io/post/go-letsencrypt/
+// Request a certificate by first registering to the ACME server and then by
+// completing a challenge then it returns the certificate as string type
+func getCert(wwwDir string, certDir string, domain string) (string, error) {
+
+	certPath := certDir + "/" + domain
```

Comment:
This might not work on Windows, consider using `path.Join` from the path package.